### ISSUE TITLE
Long processing time when output voc size large

### ISSUE THREAD

**Issue by zxu7** [datalogue/keras-attention#11]
I tried to use this layer in a machine translation task. I set the output_dim in your Attention layer as the output voc size (10k+). It seems to take the network forever to compile on my laptop. Any advice? Thanks!

**Comment by zafarali** [comment]
Hi there!

Thanks for trying it out. My implementation can be made slightly more
efficient if I incorporated the inbuilt LSTM implementation (they use one
matrix for all the weights of all the gates). In general, attention
mechanisms are bulky to train compared to regular LSTMs. Another possible
speed boost can be in hierarchical softmaxes which can be used to handle
large vocabulary sizes.

The other obvious alternative is to get a GPU! I’ve only compiled small
models on my laptop, when I needed to go bigger I went for a GPU

What is your sequence length?

**Comment by zxu7** [comment]
Thank you for responding!

I managed to run on a small set of data (8k vocab size, 8000 sequences) on a GPU now. However the code returns MemoryError when I give it the full data (200,000 sequences). 

I found that there are approaches which only take high frequency words and replace the low freq ones with <unk>. I am worried it might hurt performance and would like to hear your thoughts.

I haven't looked into/applied hierarchical softmax

**Comment by zafarali** [comment]
Instead of loading the full data into memory, perhaps you can try using
generators to load the data bit by bit? Use the keras.fit_generator method.
To do this you’d need to have your data stored in separate files or in
something like an hdf5.

**Comment by mw66** [comment]
@zafarali 

one question: "... when I needed to go bigger I went for a GPU"

Does the current `AttentionDecoder` (https://github.com/datalogue/keras-attention/blob/master/models/custom_recurrents.py) run on GPU at all?

I tried some small training example, but didn't see any GPU usage via `nvidia-smi`, the `GPU-Util` is always 0.

**Comment by mw66** [comment]
@zxu7 

"I managed to run on a small set of data (8k vocab size, 8000 sequences) on a GPU now. "

How to make this code run on GPU? can you create a PR?

Thanks.
